gcc/rust/typecheck/rust-tyty-coercion.h

```diff
@@ -1087,6 +1110,75 @@ class ReferenceCoercionRules : public BaseCoercionRules
   ReferenceType *base;
 };
 
+class PointerCoercionRules : public BaseCoercionRules
+{
+  using Rust::TyTy::BaseCoercionRules::visit;
+
+public:
+  PointerCoercionRules (PointerType *base)
+    : BaseCoercionRules (base), base (base)
+  {}
+
+  void visit (ReferenceType &type) override
+  {
+    auto base_type = base->get_base ();
+    auto other_base_type = type.get_base ();
+
+    TyTy::BaseType *base_resolved = base_type->unify (other_base_type);
+    if (base_resolved == nullptr
+	|| base_resolved->get_kind () == TypeKind::ERROR)
+      {
+	BaseCoercionRules::visit (type);
+	return;
+      }
+
+    // we can allow for mutability changes here by casting down from mutability
+    // eg:  mut vs const, we cant take a mutable reference from a const
+    // eg:  const vs mut we can take a const reference from a mutable one
+    if (!base->is_mutable () || (base->is_mutable () == type.is_mutable ()))
+      {
+	resolved = new PointerType (base->get_ref (), base->get_ty_ref (),
+				    TyVar (base_resolved->get_ref ()),
+				    base->is_mutable ());
+	return;
+      }
+
+    BaseCoercionRules::visit (type);
+  }
+
+  void visit (PointerType &type) override
+  {
+    auto base_type = base->get_base ();
+    auto other_base_type = type.get_base ();
+
+    TyTy::BaseType *base_resolved = base_type->unify (other_base_type);
+    if (base_resolved == nullptr
+	|| base_resolved->get_kind () == TypeKind::ERROR)
+      {
+	BaseCoercionRules::visit (type);
+	return;
+      }
+
+    // we can allow for mutability changes here by casting down from mutability
+    // eg:  mut vs const, we cant take a mutable reference from a const
+    // eg:  const vs mut we can take a const reference from a mutable one
```

Comment:
In this comment: s/reference/raw pointer

Reply:
fixed